### ISSUE TITLE
Losing connection flutter Firestore

### ISSUE THREAD

**Issue by nanozu** [FirebaseExtended/flutterfire#2523]
hi 
I developped a Quizz app with flutter and i use firestore for the backend, my problem is the app crashe so many times in the emulator ; when i builded the APK and i installed it in my tablet for test, the problem still, it works for some times but it crashe.

**Flutter Doctor**
Doctor summary (to see all details, run flutter doctor -v):
[√] Flutter (Channel stable, v1.12.13+hotfix.8, on Microsoft Windows [version 6.3.9600], locale fr-FR)
[√] Android toolchain - develop for Android devices (Android SDK version 28.0.3)
[√] Android Studio (version 3.5)
[√] VS Code (version 1.41.0)
[√] Connected device (1 available)

• No issues found!

**pubspec.yaml**
dev_dependencies:
firebase_core: ^0.4.4+3
firebase_auth: ^0.16.0
cloud_firestore: ^0.13.5
http: ^0.12.0+4

**app/build.gradle**
implementation : com.google.firebase:firebase-analytics:17.4.0
minSdkVersion 20
targetSdkVersion 28
multiDexEnabled true

**build.gradle**
classpath com.google.gms:google-services:4.3.3
classpath com.android.tools.build:gradle:3.5.3

**The error** 

```
E/CloudFirestorePlugin( 9605): Timed out waiting for Task
E/CloudFirestorePlugin( 9605): java.util.concurrent.TimeoutException: Timed out waiting for Task
E/CloudFirestorePlugin( 9605): 	at com.google.android.gms.tasks.Tasks.await(Unknown Source:32)
E/CloudFirestorePlugin( 9605): 	at io.flutter.plugins.firebase.cloudfirestore.CloudFirestorePlugin$4.apply(CloudFirestorePlugin.java:571)
E/CloudFirestorePlugin( 9605): 	at io.flutter.plugins.firebase.cloudfirestore.CloudFirestorePlugin$4.apply(CloudFirestorePlugin.java:523)
E/CloudFirestorePlugin( 9605): 	at com.google.firebase.firestore.FirebaseFirestore.lambda$runTransaction$0(com.google.firebase:firebase-firestore@@21.3.0:301)
E/CloudFirestorePlugin( 9605): 	at com.google.firebase.firestore.FirebaseFirestore$$Lambda$5.call(Unknown Source:6)
E/CloudFirestorePlugin( 9605): 	at com.google.android.gms.tasks.zzv.run(Unknown Source:2)
E/CloudFirestorePlugin( 9605): 	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1167)
E/CloudFirestorePlugin( 9605): 	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:641)
E/CloudFirestorePlugin( 9605): 	at java.lang.Thread.run(Thread.java:764)
E/flutter ( 9605): [ERROR:flutter/lib/ui/ui_dart_state.cc(157)] Unhandled Exception: PlatformException(Error performing transaction, Every document read in a transaction must also be written., null)
E/flutter ( 9605): #0      StandardMethodCodec.decodeEnvelope (package:flutter/src/services/message_codecs.dart:569:7)
E/flutter ( 9605): #1      MethodChannel.invokeMethod (package:flutter/src/services/platform_channel.dart:321:33)
E/flutter ( 9605): <asynchronous suspension>
E/flutter ( 9605): #2      MethodChannel.invokeMapMethod (package:flutter/src/services/platform_channel.dart:349:48)
E/flutter ( 9605): #3      MethodChannelFirestore.runTransaction (package:cloud_firestore_platform_interface/src/method_channel/method_channel_firestore.dart:123:10)
E/flutter ( 9605): #4      Firestore.runTransaction (package:cloud_firestore/src/firestore.dart:85:22)
E/flutter ( 9605): #5      _question41_bisState.build.<anonymous closure>.<anonymous closure>.<anonymous closure> (package:survey_webapp/pages/questions/question41_bis.dart:149:32)
E/flutter ( 9605): #6      _InkResponseState._handleTap (package:flutter/src/material/ink_well.dart:706:14)
E/flutter ( 9605): #7      _InkResponseState.build.<anonymous closure> (package:flutter/src/material/ink_well.dart:789:36)
E/flutter ( 9605): #8      GestureRecognizer.invokeCallback (package:flutter/src/gestures/recognizer.dart:182:24)
E/flutter ( 9605): #9      TapGestureRecognizer.handleTapUp (package:flutter/src/gestures/tap.dart:486:11)
E/flutter ( 9605): #10     BaseTapGestureRecognizer._checkUp (package:flutter/src/gestures/tap.dart:264:5)
E/flutter ( 9605): #11     BaseTapGestureRecognizer.acceptGesture (package:flutter/src/gestures/tap.dart:236:7)
E/flutter ( 9605): #12     GestureArenaManager.sweep (package:flutter/src/gestures/arena.dart:156:27)
E/flutter ( 9605): #13     GestureBinding.handleEvent (package:flutter/src/gestures/binding.dart:222:20)
E/flutter ( 9605): #14     GestureBinding.dispatchEvent (package:flutter/src/gestures/binding.dart:198:22)
E/flutter ( 9605): #15     GestureBinding._handlePointerEvent (package:flutter/src/gestures/binding.dart:156:7)
E/flutter ( 9605): #16     GestureBinding._flushPointerEventQueue (package:flutter/src/gestures/binding.dart:102:7)
E/flutter ( 9605): #17     GestureBinding._handlePointerDataPacket (package:flutter/src/gestures/binding.dart:86:7)
E/flutter ( 9605): #18     _rootRunUnary (dart:async/zone.dart:1138:13)
E/flutter ( 9605): #19     _CustomZone.runUnary (dart:async/zone.dart:1031:19)
E/flutter ( 9605): #20     _CustomZone.runUnaryGuarded (dart:async/zone.dart:933:7)
E/flutter ( 9605): #21     _invoke1 (dart:ui/hooks.dart:273:10)
E/flutter ( 9605): #22     _dispatchPointerDataPacket (dart:ui/hooks.dart:182:5)
E/flutter ( 9605): 
E/CloudFirestorePlugin( 9605): Timed out waiting for Task
E/CloudFirestorePlugin( 9605): java.util.concurrent.TimeoutException: Timed out waiting for Task
E/CloudFirestorePlugin( 9605): 	at com.google.android.gms.tasks.Tasks.await(Unknown Source:32)
E/CloudFirestorePlugin( 9605): 	at io.flutter.plugins.firebase.cloudfirestore.CloudFirestorePlugin$4.apply(CloudFirestorePlugin.java:571)
E/CloudFirestorePlugin( 9605): 	at io.flutter.plugins.firebase.cloudfirestore.CloudFirestorePlugin$4.apply(CloudFirestorePlugin.java:523)
E/CloudFirestorePlugin( 9605): 	at com.google.firebase.firestore.FirebaseFirestore.lambda$runTransaction$0(com.google.firebase:firebase-firestore@@21.3.0:301)
E/CloudFirestorePlugin( 9605): 	at com.google.firebase.firestore.FirebaseFirestore$$Lambda$5.call(Unknown Source:6)
E/CloudFirestorePlugin( 9605): 	at com.google.android.gms.tasks.zzv.run(Unknown Source:2)
E/CloudFirestorePlugin( 9605): 	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1167)
E/CloudFirestorePlugin( 9605): 	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:641)
E/CloudFirestorePlugin( 9605): 	at java.lang.Thread.run(Thread.java:764)
E/flutter ( 9605): [ERROR:flutter/lib/ui/ui_dart_state.cc(157)] Unhandled Exception: PlatformException(Error performing transaction, Every document read in a transaction must also be written., null)
E/flutter ( 9605): #0      StandardMethodCodec.decodeEnvelope (package:flutter/src/services/message_codecs.dart:569:7)
E/flutter ( 9605): #1      MethodChannel.invokeMethod (package:flutter/src/services/platform_channel.dart:321:33)
E/flutter ( 9605): <asynchronous suspension>
E/flutter ( 9605): #2      MethodChannel.invokeMapMethod (package:flutter/src/services/platform_channel.dart:349:48)
E/flutter ( 9605): #3      MethodChannelFirestore.runTransaction (package:cloud_firestore_platform_interface/src/method_channel/method_channel_firestore.dart:123:10)
E/flutter ( 9605): #4      Firestore.runTransaction (package:cloud_firestore/src/firestore.dart:85:22)
E/flutter ( 9605): #5      _question5State.build.<anonymous closure>.<anonymous closure>.<anonymous closure> (package:survey_webapp/pages/questions/question5.dart:135:50)
E/flutter ( 9605): #6      _InkResponseState._handleTap (package:flutter/src/material/ink_well.dart:706:14)
E/flutter ( 9605): #7      _InkResponseState.build.<anonymous closure> (package:flutter/src/material/ink_well.dart:789:36)
E/flutter ( 9605): #8      GestureRecognizer.invokeCallback (package:flutter/src/gestures/recognizer.dart:182:24)
E/flutter ( 9605): #9      TapGestureRecognizer.handleTapUp (package:flutter/src/gestures/tap.dart:486:11)
E/flutter ( 9605): #10     BaseTapGestureRecognizer._checkUp (package:flutter/src/gestures/tap.dart:264:5)
E/flutter ( 9605): #11     BaseTapGestureRecognizer.handlePrimaryPointer (package:flutter/src/gestures/tap.dart:199:7)
E/flutter ( 9605): #12     PrimaryPointerGestureRecognizer.handleEvent (package:flutter/src/gestures/recognizer.dart:467:9)
E/flutter ( 9605): #13     PointerRouter._dispatch (package:flutter/src/gestures/pointer_router.dart:76:12)
E/flutter ( 9605): #14     PointerRouter._dispatchEventToRoutes.<anonymous closure> (package:flutter/src/gestures/pointer_router.dart:117:9)
E/flutter ( 9605): #15     _LinkedHashMapMixin.forEach (dart:collection-patch/compact_hash.dart:379:8)
E/flutter ( 9605): #16     PointerRouter._dispatchEventToRoutes (package:flutter/src/gestures/pointer_router.dart:115:18)
E/flutter ( 9605): #17     PointerRouter.route (package:flutter/src/gestures/pointer_router.dart:101:7)
E/flutter ( 9605): #18     GestureBinding.handleEvent (package:flutter/src/gestures/binding.dart:218:19)
E/flutter ( 9605): #19     GestureBinding.dispatchEvent (package:flutter/src/gestures/binding.dart:198:22)
E/flutter ( 9605): #20     GestureBinding._handlePointerEvent (package:flutter/src/gestures/binding.dart:156:7)
E/flutter ( 9605): #21     GestureBinding._flushPointerEventQueue (package:flutter/src/gestures/binding.dart:102:7)
E/flutter ( 9605): #22     GestureBinding._handlePointerDataPacket (package:flutter/src/gestures/binding.dart:86:7)
E/flutter ( 9605): #23     _rootRunUnary (dart:async/zone.dart:1138:13)
E/flutter ( 9605): #24     _CustomZone.runUnary (dart:async/zone.dart:1031:19)
E/flutter ( 9605): #25     _CustomZone.runUnaryGuarded (dart:async/zone.dart:933:7)
E/flutter ( 9605): #26     _invoke1 (dart:ui/hooks.dart:273:10)
E/flutter ( 9605): #27     _dispatchPointerDataPacket (dart:ui/hooks.dart:182:5)
E/flutter ( 9605): 
I/flutter ( 9605): task updated
E/CloudFirestorePlugin( 9605): Timed out waiting for Task
E/CloudFirestorePlugin( 9605): java.util.concurrent.TimeoutException: Timed out waiting for Task
E/CloudFirestorePlugin( 9605): 	at com.google.android.gms.tasks.Tasks.await(Unknown Source:32)
E/CloudFirestorePlugin( 9605): 	at io.flutter.plugins.firebase.cloudfirestore.CloudFirestorePlugin$4.apply(CloudFirestorePlugin.java:571)
E/CloudFirestorePlugin( 9605): 	at io.flutter.plugins.firebase.cloudfirestore.CloudFirestorePlugin$4.apply(CloudFirestorePlugin.java:523)
E/CloudFirestorePlugin( 9605): 	at com.google.firebase.firestore.FirebaseFirestore.lambda$runTransaction$0(com.google.firebase:firebase-firestore@@21.3.0:301)
E/CloudFirestorePlugin( 9605): 	at com.google.firebase.firestore.FirebaseFirestore$$Lambda$5.call(Unknown Source:6)
E/CloudFirestorePlugin( 9605): 	at com.google.android.gms.tasks.zzv.run(Unknown Source:2)
E/CloudFirestorePlugin( 9605): 	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1167)
E/CloudFirestorePlugin( 9605): 	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:641)
E/CloudFirestorePlugin( 9605): 	at java.lang.Thread.run(Thread.java:764)
E/flutter ( 9605): [ERROR:flutter/lib/ui/ui_dart_state.cc(157)] Unhandled Exception: PlatformException(Error performing transaction, Every document read in a transaction must also be written., null)
E/flutter ( 9605): #0      StandardMethodCodec.decodeEnvelope (package:flutter/src/services/message_codecs.dart:569:7)
E/flutter ( 9605): #1      MethodChannel.invokeMethod (package:flutter/src/services/platform_channel.dart:321:33)
E/flutter ( 9605): <asynchronous suspension>
E/flutter ( 9605): #2      MethodChannel.invokeMapMethod (package:flutter/src/services/platform_channel.dart:349:48)
E/flutter ( 9605): #3      MethodChannelFirestore.runTransaction (package:cloud_firestore_platform_interface/src/method_channel/method_channel_firestore.dart:123:10)
E/flutter ( 9605): #4      Firestore.runTransaction (package:cloud_firestore/src/firestore.dart:85:22)
E/flutter ( 9605): #5      _question1State.build.<anonymous closure>.<anonymous closure>.<anonymous closure> (package:survey_webapp/pages/questions/question1.dart:143:50)
E/flutter ( 9605): #6      _InkResponseState._handleTap (package:flutter/src/material/ink_well.dart:706:14)
E/flutter ( 9605): #7      _InkResponseState.build.<anonymous closure> (package:flutter/src/material/ink_well.dart:789:36)
E/flutter ( 9605): #8      GestureRecognizer.invokeCallback (package:flutter/src/gestures/recognizer.dart:182:24)
E/flutter ( 9605): #9      TapGestureRecognizer.handleTapUp (package:flutter/src/gestures/tap.dart:486:11)
E/flutter ( 9605): #10     BaseTapGestureRecognizer._checkUp (package:flutter/src/gestures/tap.dart:264:5)
E/flutter ( 9605): #11     BaseTapGestureRecognizer.handlePrimaryPointer (package:flutter/src/gestures/tap.dart:199:7)
E/flutter ( 9605): #12     PrimaryPointerGestureRecognizer.handleEvent (package:flutter/src/gestures/recognizer.dart:467:9)
E/flutter ( 9605): #13     PointerRouter._dispatch (package:flutter/src/gestures/pointer_router.dart:76:12)
E/flutter ( 9605): #14     PointerRouter._dispatchEventToRoutes.<anonymous closure> (package:flutter/src/gestures/pointer_router.dart:117:9)
E/flutter ( 9605): #15     _LinkedHashMapMixin.forEach (dart:collection-patch/compact_hash.dart:379:8)
E/flutter ( 9605): #16     PointerRouter._dispatchEventToRoutes (package:flutter/src/gestures/pointer_router.dart:115:18)
E/flutter ( 9605): #17     PointerRouter.route (package:flutter/src/gestures/pointer_router.dart:101:7)
E/flutter ( 9605): #18     GestureBinding.handleEvent (package:flutter/src/gestures/binding.dart:218:19)
E/flutter ( 9605): #19     GestureBinding.dispatchEvent (package:flutter/src/gestures/binding.dart:198:22)
E/flutter ( 9605): #20     GestureBinding._handlePointerEvent (package:flutter/src/gestures/binding.dart:156:7)
E/flutter ( 9605): #21     GestureBinding._flushPointerEventQueue (package:flutter/src/gestures/binding.dart:102:7)
E/flutter ( 9605): #22     GestureBinding._handlePointerDataPacket (package:flutter/src/gestures/binding.dart:86:7)
E/flutter ( 9605): #23     _rootRunUnary (dart:async/zone.dart:1138:13)
E/flutter ( 9605): #24     _CustomZone.runUnary (dart:async/zone.dart:1031:19)
E/flutter ( 9605): #25     _CustomZone.runUnaryGuarded (dart:async/zone.dart:933:7)
E/flutter ( 9605): #26     _invoke1 (dart:ui/hooks.dart:273:10)
E/flutter ( 9605): #27     _dispatchPointerDataPacket (dart:ui/hooks.dart:182:5)
E/flutter ( 9605): 
E/CloudFirestorePlugin( 9605): Timed out waiting for Task
E/CloudFirestorePlugin( 9605): java.util.concurrent.TimeoutException: Timed out waiting for Task
E/CloudFirestorePlugin( 9605): 	at com.google.android.gms.tasks.Tasks.await(Unknown Source:32)
E/CloudFirestorePlugin( 9605): 	at io.flutter.plugins.firebase.cloudfirestore.CloudFirestorePlugin$4.apply(CloudFirestorePlugin.java:571)
E/CloudFirestorePlugin( 9605): 	at io.flutter.plugins.firebase.cloudfirestore.CloudFirestorePlugin$4.apply(CloudFirestorePlugin.java:523)
E/CloudFirestorePlugin( 9605): 	at com.google.firebase.firestore.FirebaseFirestore.lambda$runTransaction$0(com.google.firebase:firebase-firestore@@21.3.0:301)
E/CloudFirestorePlugin( 9605): 	at com.google.firebase.firestore.FirebaseFirestore$$Lambda$5.call(Unknown Source:6)
E/CloudFirestorePlugin( 9605): 	at com.google.android.gms.tasks.zzv.run(Unknown Source:2)
E/CloudFirestorePlugin( 9605): 	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1167)
E/CloudFirestorePlugin( 9605): 	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:641)
E/CloudFirestorePlugin( 9605): 	at java.lang.Thread.run(Thread.java:764)
E/CloudFirestorePlugin( 9605): Timed out waiting for Task
E/CloudFirestorePlugin( 9605): java.util.concurrent.TimeoutException: Timed out waiting for Task
E/CloudFirestorePlugin( 9605): 	at com.google.android.gms.tasks.Tasks.await(Unknown Source:32)
E/CloudFirestorePlugin( 9605): 	at io.flutter.plugins.firebase.cloudfirestore.CloudFirestorePlugin$4.apply(CloudFirestorePlugin.java:571)
E/CloudFirestorePlugin( 9605): 	at io.flutter.plugins.firebase.cloudfirestore.CloudFirestorePlugin$4.apply(CloudFirestorePlugin.java:523)
E/CloudFirestorePlugin( 9605): 	at com.google.firebase.firestore.FirebaseFirestore.lambda$runTransaction$0(com.google.firebase:firebase-firestore@@21.3.0:301)
E/CloudFirestorePlugin( 9605): 	at com.google.firebase.firestore.FirebaseFirestore$$Lambda$5.call(Unknown Source:6)
E/CloudFirestorePlugin( 9605): 	at com.google.android.gms.tasks.zzv.run(Unknown Source:2)
E/CloudFirestorePlugin( 9605): 	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1167)
E/CloudFirestorePlugin( 9605): 	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:641)
E/CloudFirestorePlugin( 9605): 	at java.lang.Thread.run(Thread.java:764)
E/AndroidRuntime( 9605): FATAL EXCEPTION: AsyncTask #4
E/AndroidRuntime( 9605): Process: com.surveywebapp.survey_webapp, PID: 9605
E/AndroidRuntime( 9605): java.lang.RuntimeException: An error occurred while executing doInBackground()
E/AndroidRuntime( 9605): 	at android.os.AsyncTask$3.done(AsyncTask.java:354)
E/AndroidRuntime( 9605): 	at java.util.concurrent.FutureTask.finishCompletion(FutureTask.java:383)
E/AndroidRuntime( 9605): 	at java.util.concurrent.FutureTask.setException(FutureTask.java:252)
E/AndroidRuntime( 9605): 	at java.util.concurrent.FutureTask.run(FutureTask.java:271)
E/AndroidRuntime( 9605): 	at android.os.AsyncTask$SerialExecutor$1.run(AsyncTask.java:245)
E/AndroidRuntime( 9605): 	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1167)
E/AndroidRuntime( 9605): 	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:641)
E/AndroidRuntime( 9605): 	at java.lang.Thread.run(Thread.java:764)
E/AndroidRuntime( 9605): Caused by: java.lang.AssertionError: INTERNAL ASSERTION FAILED: A transaction object cannot be used after its update callback has been invoked.
E/AndroidRuntime( 9605): 	at com.google.firebase.firestore.util.Assert.fail(com.google.firebase:firebase-firestore@@21.3.0:46)
E/AndroidRuntime( 9605): 	at com.google.firebase.firestore.util.Assert.hardAssert(com.google.firebase:firebase-firestore@@21.3.0:31)
E/AndroidRuntime( 9605): 	at com.google.firebase.firestore.core.Transaction.ensureCommitNotCalled(com.google.firebase:firebase-firestore@@21.3.0:246)
E/AndroidRuntime( 9605): 	at com.google.firebase.firestore.core.Transaction.lookup(com.google.firebase:firebase-firestore@@21.3.0:81)
E/AndroidRuntime( 9605): 	at com.google.firebase.firestore.Transaction.getAsync(com.google.firebase:firebase-firestore@@21.3.0:191)
E/AndroidRuntime( 9605): 	at com.google.firebase.firestore.Transaction.get(com.google.firebase:firebase-firestore@@21.3.0:228)
E/AndroidRuntime( 9605): 	at io.flutter.plugins.firebase.cloudfirestore.CloudFirestorePlugin$5.doInBackground(CloudFirestorePlugin.java:613)
E/AndroidRuntime( 9605): 	at io.flutter.plugins.firebase.cloudfirestore.CloudFirestorePlugin$5.doInBackground(CloudFirestorePlugin.java:608)
E/AndroidRuntime( 9605): 	at android.os.AsyncTask$2.call(AsyncTask.java:333)
E/AndroidRuntime( 9605): 	at java.util.concurrent.FutureTask.run(FutureTask.java:266)
E/AndroidRuntime( 9605): 	... 4 more
I/Process ( 9605): Sending signal. PID: 9605 SIG: 9
```

thank you !

**Comment by TahaTesser** [comment]
Hi @nanozu 
can you please provide your `flutter doctor -v` and `flutter run --verbose`?
Also, to better address the issue, would be helpful if you could post a minimal code sample to reproduce the problem
Thank you

**Comment by Salakar** [comment]
Hey this has been fixed as part of #2582 / #2913  - pending release.
